V1 documentation

cpfiffer opened the pull request:
Draft PR to implement the general structure of the v1 documentation, following issue #1528.

rlouf commented:
Are we switching to Hextra for it to look similar to the product doc? While I like the idea of having a similar design, mkdocs-material really makes a lot of things easy, such as automatically generating API reference documentation or integrating with Google Analytics. We could tweak the theme to have something similar looking, default to dark mode, etc. but I think we might be shooting ourselves in the foot by switching to Hextra.

cpfiffer commented:
It's a good point about mkdocs. Partly the goal was to have a visual refresh and to match the product documentation and present a cohesive brand image, as the Outlines docs are visually kind of messy IMO.

The advantage of sticking with mkdocs is that we have an established flow for generating the docs. It's also designed for Python and tends to work quite well.

As to Google Analytics, this is [supported in Hugo](https://gideonwolfe.com/posts/sysadmin/hugo/hugogoogleanalytics/) easily. Automatic API reference is simple as well by adding a single step before the hugo site render, where we use something like [lazydocs](https://github.com/ml-tooling/lazydocs) to generate markdown files from Python libraries.

cpfiffer commented:
I'm going to switch this back to mkdocs to limit the amount of work we'll have to do. I still want to make some visual changes though, but we may be able to do this in a series of PRs.

cpfiffer commented:
I reverted this back to the original mkdocs format, and am still updating the skeleton/moving things around.

rlouf commented on docs/index.md:
```diff
@@ -1,8 +1,154 @@
+- :material-shield-check: __Reliable__ - Guaranteed schema compliance -- always valid JSON.
+- :material-puzzle: __Feature-rich__ - Supports a large proportion of the JSON Schema spec, along with regex and context-free grammars.
+- :material-lightning-bolt: __Fast__ - Outlines has negligible runtime overhead, and fast compilation times.
+- :material-cog: __Universal__ - Outlines is a powered by Rust, and can be easily bound to other languages.
```
This is more for `outlines-core` documentation imo

cpfiffer replied:
I'll change this to say that it is extensible through a simple API, since I'd like to keep six totaly cards (and it is also true that Outlines is extensible)

rlouf commented on docs/index.md:
```diff
@@ -1,8 +1,154 @@
+- :material-lightning-bolt: __Fast__ - Outlines has negligible runtime overhead, and fast compilation times.
+- :material-cog: __Universal__ - Outlines is a powered by Rust, and can be easily bound to other languages.
+- :material-lightbulb: __Simple__ - Outlines is a low-abstraction library. Write code the way you normally do with LLMs. No agent frameworks needed.
+- :material-magnify: __Powerful__ - Manage prompt complexity with prompt templating.
```
I would put the model integrations front and center, especially with the inference servers

cpfiffer replied:
Removed the "extensible" thing and added:

> - :material-earth: __Universal__ - work with your favorite LLM inference backend, like [vLLM](https://docs.vllm.ai/en/stable/), [transformers](https://huggingface.co/docs/transformers/en/index), [OpenAI](https://platform.openai.com/docs/guides/structured-outputs), [llama.cpp](https://github.com/ggml-org/llama.cpp), [mlx-lm](https://github.com/ml-explore/mlx-lm), and more.

rlouf commented on .gitignore:
```diff
@@ -15,3 +15,4 @@ benchmarks/results
 build
 docs/build
 logs
+.cursor/
```
Can you remove this and add it to `.gitignore`?

cpfiffer replied:
I'm confused here, it's in the gitignore and I don't think I've  checked .cursor in.

rlouf commented on docs/index.md:
```diff
@@ -1,8 +1,154 @@
+    # This field uses a regex pattern to constrain the output
+    # to only valid email addresses
+    email: str = Field(
+        pattern=r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$',
```
This will fail if people change the model for OpenAI, is that what we want?

cpfiffer replied:
Hm, you're right -- probably not. Let's omit the email field entirely.

cpfiffer commented:
Note: I'm changing the structure of the docs. Is it worth it to build all the redirects we'd need for the current sitemap?

cpfiffer commented:
I have introduced many changes to the documentation, including:

- Complete reorganization, centered around the user guide, the API reference, and examples (renamed from cookbooks)
- Rewritten welcome page
- Added skeleton files for all expected documentation pages
- Simplified theming + partial light/dark mode support (it is missing syntax highlighting theme changes)

The goal of this PR is to get the basic shape down, and then to add PRs later for individual documents. I've made a few changes here and there, such as rewrite the welcome page, but for the most part this PR lays out a general structure for the docs.

I'll flag this for review.

cpfiffer commented:
Some screenshots to help visualize:
![image](https://github.com/user-attachments/assets/be110da3-8db3-422c-b7db-46bf8a6b2db8)
![image](https://github.com/user-attachments/assets/a077fb48-f18c-4ad3-a89c-a6fb00f896c2)
![image](https://github.com/user-attachments/assets/38f2fcdd-c916-4b5b-839f-b97445bf23c2)
![image](https://github.com/user-attachments/assets/33b4acaa-5984-4ea5-9958-b3efa11f15ce)

There's a _lot_ of writing to be done here as well.

cpfiffer commented:
Actually, hold off on the reviews for now. I want to do some more clean up tomorrow morning.

cpfiffer commented:
Let's merge this in early so we can iterate on it more later, @RobinPicard @rlouf

RobinPicard commented:
Awesome, I love the clean look of the website! Thanks so much for the hard work @cpfiffer. I'm in favor of merging it right away as we will update some of the sections in other PRs anyway.

RobinPicard reviewed (approved):
Awesome, I love the clean look of the website! Thanks so much for the hard work @cpfiffer. I'm in favor of merging it right away as we will update some of the sections in other PRs anyway.

rlouf commented:
Just run pre-commit to remove the trailing spaces, and let's merge this!

cpfiffer commented:
Ran pre-commit, should be good.
